File: src-native/ios/CartoAdditions/AKHTTPTileDataSource.m
//
//  AKHTTPTileDataSource.m
//  CartoAdditions
//
//  Created by Martin Guillon on 3/8/19.
//  Copyright © 2019 Stefan Dragnev. All rights reserved.
//

#import "AKHTTPTileDataSource.h"

@implementation AKHTTPTileDataSource
{
   BOOL mTileSizePixels;
   BOOL mEnableSSL;
  BOOL isVector ;
  CGFloat mDpi ;
  CGFloat mOpacity;
   BOOL mAutoHD ;
   BOOL mVisible;
  CGFloat mMinimumZoomLevel ;
  CGFloat mMaximumZoomLevel;
}

-(id)init {
  if (self = [super init]) {
    mTileSizePixels = 256.0;
    mEnableSSL = false;
    isVector = false;
    mDpi = 1.0;
    mOpacity = 1.0;
    mAutoHD = false;
    mVisible = true;
    mMinimumZoomLevel = -1.0;
    mMaximumZoomLevel = -1.0;
  }
  return self;
}

-(id)initWithMinZoom: (int)minZoom maxZoom: (int)maxZoom baseURL: (NSString*)baseURL {
  if (self = [super initWithMinZoom:minZoom maxZoom:maxZoom baseURL:baseURL]) {
    isVector = [baseURL containsString:@"mvt"] || [baseURL containsString:@"pbf"] || [baseURL hasPrefix:@"carto."];
  }
  return self;
}

-(id) setVisible:(BOOL)value  {
  mVisible = value;
  return self;
}

-(id) setAutoHD:(BOOL)value  {
  mAutoHD = value;
  return self;
}
-(id) setMinimumZoomLevel:(CGFloat)value  {
  mMinimumZoomLevel = value;
  return self;
}
-(id) setMaximumZoomLevel:(CGFloat)value  {
  mMinimumZoomLevel = value;
  return self;
}

-(id) setTileSize:(CGFloat)value  {
  mTileSizePixels = value;
  return self;
}
-(id) setOpacity:(CGFloat)value  {
  mOpacity = value;
  return self;
}

-(CGFloat) getMinimumZoomLevel {
  return mMinimumZoomLevel;
}

-(CGFloat) getMaximumZoomLeve  {
  return mMaximumZoomLevel;
}

-(CGFloat) getTileSizePixel {
  return mTileSizePixels;
}
-(CGFloat) getCenterZoom  {
  return roundf(mMaximumZoomLevel + mMinimumZoomLevel) / 2;
}

-(NTTileData*) loadTile:(NTMapTile*)mapTile {
  BOOL needsHd = !isVector && mAutoHD && (mTileSizePixels / mDpi) <= 128 ;
  double x = [mapTile getX];
  double y = [mapTile getY];
  double z = [mapTile getZoom];
  double hdZ = (needsHd ? (z + 1) : z);
  BOOL passedMin = !isVector && mMinimumZoomLevel >= 0 && (hdZ < mMinimumZoomLevel);
  
  //    if (passedMin && !mShowTileBeforeMaxZoom) {
  //      return null;
  //    }
  
  BOOL passedMax = !isVector && mMaximumZoomLevel >= 0 && hdZ > mMaximumZoomLevel;
  //    if (passedMax && !mShowTileAfterMaxZoom) {
  //      return null;
  //    }
  
  //        if (mVisible == false || mOpacity == 0.0f) {
  //            return null;
  //        }
  NTTileData* tileData = NULL;
  BOOL needsCrop = false;
  double deltaCropX = 0.0;
  double deltaCropY = 0.0;
  double cropSize = 0.0;
  if (passedMin) {
    // needsCrop = true;
    // int minZoom = (int) (needsHd ? (mMinimumZoomLevel - 1)
    // : mMinimumZoomLevel);
    // float currentTileDepth = mMinimumZoomLevel - z;
    // double nextx = x / Math.pow(2.0, currentTileDepth);
    // double nexty = y / Math.pow(2.0, currentTileDepth);
    // x = (int) Math.floor(nextx);
    // y = (int) Math.floor(nexty);
    // z = maxZoom;
    // cropSize = 1.0f / Math.pow(2.0, currentTileDepth);
    // deltaCropX = nextx - x;
    // deltaCropY = nexty - y;
  } else if (passedMax) {
    needsCrop = true;
    CGFloat maxZoom = (needsHd ? (mMaximumZoomLevel - 1)
                   : mMaximumZoomLevel);
    double currentTileDepth = z - maxZoom;
    double nextx = x / pow(2.0, currentTileDepth);
    double nexty = y / pow(2.0, currentTileDepth);
    x = floor(nextx);
    y = floor(nexty);
    z = maxZoom;
    cropSize = 1.0 / pow(2.0, currentTileDepth);
    deltaCropX = nextx - x;
    deltaCropY = nexty - y;
  }
  if (needsHd) {
    tileData = [self getTileFromNextZoomLevel:x y:y z:z];
  } else {
    tileData = [super loadTile:mapTile];
  }
  if (tileData != nil) {
    if (needsCrop) {
      UIImage* image  = [self getUIImage:[tileData getData]];
      if (image) {
        double width = image.size.width;
        double height = image.size.height;
        image = [self crop
                 :image cropRect:CGRectMake( (int)(width * deltaCropX) , (int)(height * deltaCropY), (int)(width * cropSize), (int)(height * cropSize))];
      }
      
      
      //        var data = tileData.getData().getData();
      //        Bitmap tileImage = BitmapFactory.decodeByteArray(data, 0, data.length, null);
      //        if (tileImage != null) {
      //          Bitmap oldBitmap = tileImage;
      //          final int width = tileImage.getWidth();
      //          final int height = tileImage.getHeight();
      //          tileImage = Bitmap.createBitmap(tileImage,
      //                                          (int) (width * deltaCropX), (int) (height * deltaCropY),
      //                                          (int) (width * cropSize), (int) (height * cropSize));
      //          if (oldBitmap != tileImage) {
      //            oldBitmap.recycle();
      //          }
      //          return new TileData(BitmapUtils.createBitmapFromAndroidBitmap(tileImage).getPixelData());
    } else {
      return nil;
    }
    
  }
  
  return tileData;
}


-(UIImage*) getUIImage:(NTBinaryData*)tileData {
  unsigned char * data = [tileData getData];
  NSUInteger length = [tileData size];
  
  NSData* nsData = [NSData dataWithBytes:data length:length];
  return [UIImage imageWithData:nsData];
}

-(NTBinaryData*) getNTBinaryData:(UIImage*)image {
  NSData*  imageData = UIImagePNGRepresentation(image) ;
  NSUInteger length = [imageData length];
  unsigned char *bytePtr = (unsigned char *)[imageData bytes];
  return [[NTBinaryData alloc] initWithDataPtr:bytePtr size:(int)length];
}


-(NTTileData*) mergeTileDatas:(NTTileData*)tile1 tile2:(NTTileData*)tile2 tile3:(NTTileData*)tile3 tile4:(NTTileData*)tile4 {
  if (tile1 == nil || tile2 == nil || tile3 == nil || tile4 == nil) {
    return nil;
  }
  CGSize size = CGSizeMake(512, 512);
  UIGraphicsBeginImageContext(size);
  
  //    let areaSize = CGRect(x: 0, y: 0, width: size.width, height: size.height)
  
  NTBinaryData* tileData = [tile1 getData];
  if (tileData != nil) {
    UIImage* image  = [self getUIImage:tileData];
    [image drawAtPoint:CGPointMake(0,0)];
//    image?.draw(at: CGPoint(x:0,y:0));
  }
  
  
  UIImage* newImage = UIGraphicsGetImageFromCurrentImageContext();
  UIGraphicsEndImageContext();
  return [[NTTileData alloc ] initWithData:[self getNTBinaryData:newImage]];
  
}

-(NTTileData*) getTileFromNextZoomLevel:(int)x y:(int)y z:(int)z {
//  let group = DispatchGroup()
//  group.enter()
  __block NTTileData* result = nil;
  dispatch_queue_t serialQ = dispatch_queue_create("serialQ", DISPATCH_QUEUE_SERIAL);
  dispatch_async(serialQ, ^
  {
    NTTileData* data1 = [super loadTile:[[NTMapTile alloc] initWithX:x * 2 y:y * 2 zoom:z + 1 frameNr:0]];
    NTTileData* data2 = [super loadTile:[[NTMapTile alloc] initWithX:x * 2 + 1 y:y * 2 zoom:z + 1 frameNr:0]];
    NTTileData* data3 = [super loadTile:[[NTMapTile alloc] initWithX:x * 2  y:y * 2 + 1 zoom:z + 1 frameNr:0]];
    NTTileData* data4 = [super loadTile:[[NTMapTile alloc] initWithX:x * 2 + 1 y:y * 2 + 1 zoom:z + 1 frameNr:0]];

    result = [self mergeTileDatas: data1 tile2: data2 tile3: data3 tile4: data4];
  });
  dispatch_sync(serialQ, ^{ });
//dispatch_async(dispatch_get_global_queue( DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
//    group.leave()
//  })
//  group.wait()
  return result;
}

-(UIImage*) crop: (UIImage*)image  cropRect:(CGRect)cropRect {
  UIGraphicsBeginImageContextWithOptions(cropRect.size, false, image.scale);
  CGPoint origin = CGPointMake( cropRect.origin.x * -1.0,  cropRect.origin.y * -1.0);
  [image drawAtPoint:origin];
  UIImage* result = UIGraphicsGetImageFromCurrentImageContext();
  UIGraphicsEndImageContext();
  
  return result;
}
@end

File: src-native/ios/CartoAdditions/AKHTTPTileDataSource.swift
//
//  HTTPTileDataSource.swift
//  CartoAdditions
//
//  Created by Martin Guillon on 10/3/18.
//  Copyright © 2018 Telerik A D. All rights reserved.
//

import Foundation


extension StringProtocol where Index == String.Index {
  func index(of string: Self, options: String.CompareOptions = []) -> Index? {
    return range(of: string, options: options)?.lowerBound
  }
  func endIndex(of string: Self, options: String.CompareOptions = []) -> Index? {
    return range(of: string, options: options)?.upperBound
  }
  func indexes(of string: Self, options: String.CompareOptions = []) -> [Index] {
    var result: [Index] = []
    var start = startIndex
    while start < endIndex,
      let range = self[start..<endIndex].range(of: string, options: options) {
        result.append(range.lowerBound)
        start = range.lowerBound < range.upperBound ? range.upperBound :
          index(range.lowerBound, offsetBy: 1, limitedBy: endIndex) ?? endIndex
    }
    return result
  }
  func ranges(of string: Self, options: String.CompareOptions = []) -> [Range<Index>] {
    var result: [Range<Index>] = []
    var start = startIndex
    while start < endIndex,
      let range = self[start..<endIndex].range(of: string, options: options) {
        result.append(range)
        start = range.lowerBound < range.upperBound ? range.upperBound :
          index(range.lowerBound, offsetBy: 1, limitedBy: endIndex) ?? endIndex
    }
    return result
  }
}

class AKHTTPTileDataSource: NTHTTPTileDataSource {
  private var mTileSizePixels = 256.0;
  var mEnableSSL = false;
  var isVector = false;
  var mDpi = 1.0;
  private var mOpacity = 1.0;
  private var mAutoHD = false;
  private var mVisible = true;
  var mMinimumZoomLevel = -1.0;
  var mMaximumZoomLevel = -1.0;
  
  override init!(minZoom: Int32, maxZoom: Int32, baseURL: String!) {
    
    super.init(minZoom: minZoom, maxZoom: maxZoom, baseURL: baseURL)
    self.isVector = baseURL.range(of:"mvt") != nil || baseURL.range(of:"pbf") != nil || baseURL.hasPrefix("carto.");
  }
  
  func setVisible(visible: Bool)->AKHTTPTileDataSource  {
    self.mVisible = visible;
    return self;
  }
  
  func setAutoHD(hd: Bool)->AKHTTPTileDataSource {
    self.mAutoHD = hd;
    return self;
  }

  
  func setMinimumZoomLevel(aMinimumZoomLevel: Double)->AKHTTPTileDataSource {
    self.mMinimumZoomLevel = aMinimumZoomLevel;
    return self;
  }
  
  /**
   * Sets the layer's minimum zoom level.
   */
  func setMaximumZoomLevel(aMaximumZoomLevel: Double)->AKHTTPTileDataSource {
    self.mMaximumZoomLevel = aMaximumZoomLevel;
    return self;
  }
  
  func setTileSize(size: Double)->AKHTTPTileDataSource {
    self.mTileSizePixels = size;
    return self;
  }
  
  func setOpacity(opacity:Double)->AKHTTPTileDataSource {
    self.mOpacity = opacity;
    return self;
  }
  
  func getMinimumZoomLevel()->Double {
    return self.mMinimumZoomLevel;
  }
  
  func getMaximumZoomLevel()->Double  {
    return self.mMaximumZoomLevel;
  }
  
  func getTileSizePixels() ->Double {
    return mTileSizePixels;
  }
  
  func getCenterZoom()->Double  {
    return round(mMaximumZoomLevel + mMinimumZoomLevel) / 2;
  }
  
  override func load(_ mapTile: NTMapTile!) -> NTTileData! {
    let needsHd = !isVector && mAutoHD && (mTileSizePixels / mDpi) <= 128 ;
    var x = Double(mapTile.getX());
    var y = Double(mapTile.getY());
    var z = Double(mapTile.getZoom());
    let hdZ = (needsHd ? (z + 1) : z);
    let passedMin = !isVector && mMinimumZoomLevel >= 0 && (hdZ < mMinimumZoomLevel);
    
//    if (passedMin && !mShowTileBeforeMaxZoom) {
//      return null;
//    }
    
    let passedMax = !isVector && mMaximumZoomLevel >= 0 && hdZ > mMaximumZoomLevel;
//    if (passedMax && !mShowTileAfterMaxZoom) {
//      return null;
//    }
    
    //        if (mVisible == false || mOpacity == 0.0f) {
    //            return null;
    //        }
    var tileData: NTTileData?;
    var needsCrop = false;
    var deltaCropX = 0.0;
    var deltaCropY = 0.0;
    var cropSize = 0.0;
    if (passedMin) {
      // needsCrop = true;
      // int minZoom = (int) (needsHd ? (mMinimumZoomLevel - 1)
      // : mMinimumZoomLevel);
      // float currentTileDepth = mMinimumZoomLevel - z;
      // double nextx = x / Math.pow(2.0, currentTileDepth);
      // double nexty = y / Math.pow(2.0, currentTileDepth);
      // x = (int) Math.floor(nextx);
      // y = (int) Math.floor(nexty);
      // z = maxZoom;
      // cropSize = 1.0f / Math.pow(2.0, currentTileDepth);
      // deltaCropX = nextx - x;
      // deltaCropY = nexty - y;
    } else if (passedMax) {
      needsCrop = true;
      let maxZoom = (needsHd ? (mMaximumZoomLevel - 1)
        : mMaximumZoomLevel);
      let currentTileDepth = Double(mapTile.getZoom()) - maxZoom;
      let nextx = Double(mapTile.getX()) / pow(2.0, currentTileDepth);
      let nexty = Double(mapTile.getY()) / pow(2.0, currentTileDepth);
      x = floor(nextx);
      y = floor(nexty);
      z = maxZoom;
      cropSize = 1.0 / pow(2.0, currentTileDepth);
      deltaCropX = nextx - x;
      deltaCropY = nexty - y;
    }
    if (needsHd) {
      tileData = getTileFromNextZoomLevel(x:Int32(x), y:Int32(y), z:Int32(z));
    } else {
      tileData = super.load(mapTile);
    }
    if (tileData != nil) {
      if (needsCrop) {
        var image  = getUIImage(tileData:tileData!.getData());
        let width = Double(image!.size.width);
        let height = Double(image!.size.height);
        image = crop(image:image!, cropRect:CGRect(x: Int(width * deltaCropX), y: Int(height * deltaCropY), width: Int(width * cropSize), height: Int(height * cropSize)));
        
//        var data = tileData.getData().getData();
//        Bitmap tileImage = BitmapFactory.decodeByteArray(data, 0, data.length, null);
//        if (tileImage != null) {
//          Bitmap oldBitmap = tileImage;
//          final int width = tileImage.getWidth();
//          final int height = tileImage.getHeight();
//          tileImage = Bitmap.createBitmap(tileImage,
//                                          (int) (width * deltaCropX), (int) (height * deltaCropY),
//                                          (int) (width * cropSize), (int) (height * cropSize));
//          if (oldBitmap != tileImage) {
//            oldBitmap.recycle();
//          }
//          return new TileData(BitmapUtils.createBitmapFromAndroidBitmap(tileImage).getPixelData());
        } else {
          return nil;
        }
        
      }
    
    return tileData;
  }
  
  
  func getUIImage(tileData:NTBinaryData)-> UIImage? {
    let data = tileData.getData();
  
    let nsData = NSData.init(bytes: data, length: Int(tileData.size()));
    return UIImage(data: nsData as Data);
  }

  func getNTBinaryData(image: UIImage?)->NTBinaryData? {
    let imageData = UIImagePNGRepresentation(image!) as NSData?;
    let length = imageData!.length;
    let uint8Ptr = UnsafeMutablePointer<UInt8>.allocate(capacity: length)
    imageData!.getBytes(uint8Ptr, length: length);
    return NTBinaryData.init(dataPtr: uint8Ptr, size: UInt32(length))
  }
  
  
  func mergeTileDatas(tile1:NTTileData?, tile2:NTTileData?, tile3:NTTileData?,tile4:NTTileData?) -> NTTileData? {
     if (tile1 == nil || tile2 == nil || tile3 == nil || tile4 == nil) {
         return nil;
     }
    let size = CGSize(width: 512, height: 512)
    UIGraphicsBeginImageContext(size)
    
//    let areaSize = CGRect(x: 0, y: 0, width: size.width, height: size.height)
    
    let tileData = tile1?.getData();
    if (tileData != nil) {
      let image  = getUIImage(tileData:tileData!);
      image?.draw(at: CGPoint(x:0,y:0));
    }
    
    
    let newImage:UIImage = UIGraphicsGetImageFromCurrentImageContext()!
    UIGraphicsEndImageContext()
    return NTTileData.init(data: getNTBinaryData(image:newImage));
  
  }
  
  func getTileFromNextZoomLevel(x: Int32, y: Int32, z: Int32)-> NTTileData? {
    let group = DispatchGroup()
    group.enter()
    var result:NTTileData?
    DispatchQueue.global(qos: .background).async {
      let data1 = super.load(NTMapTile(x:x * 2, y:y * 2, zoom:z + 1, frameNr:0));
      let data2 = super.load(NTMapTile(x:x * 2 + 1, y:y * 2, zoom:z + 1, frameNr:0));
      let data3 = super.load(NTMapTile(x:x * 2, y:y * 2 + 1, zoom:z + 1, frameNr:0));
      let data4 = super.load(NTMapTile(x:x * 2 + 1, y:y * 2 + 1, zoom:z + 1, frameNr:0));
      result = self.mergeTileDatas(tile1: data1, tile2: data2, tile3: data3, tile4: data4);
      group.leave()
    }
    group.wait()
    return result;
  }
  
  func crop(image:UIImage, cropRect:CGRect) -> UIImage? {
    UIGraphicsBeginImageContextWithOptions(cropRect.size, false, image.scale)
    let origin = CGPoint(x: cropRect.origin.x * CGFloat(-1), y: cropRect.origin.y * CGFloat(-1))
    image.draw(at: origin)
    let result = UIGraphicsGetImageFromCurrentImageContext()
    UIGraphicsEndImageContext();
    
    return result
  }

}
